File: Solutions/python/pig-latin/pig_latin.py
def translate(text):
    vowels = "aeiou"
    words = text.split()

    def pig_latin(word):
        if word[0] in vowels or word.startswith("xr") or word.startswith("yt"):
            return word + "ay"

        for i in range(len(word)):
            if word[i] in vowels:
                if word[i - 1 : i + 1] == "qu":
                    return word[i + 1 :] + word[: i + 1] + "ay"
                return word[i:] + word[:i] + "ay"
            if word[i] == "y" and i != 0:
                return word[i:] + word[:i] + "ay"

        return word[1:] + word[0] + "ay" if word[0] == "y" else word + "ay"

    return " ".join([pig_latin(word) for word in words])

File: Solutions/elixir/pig-latin/lib/pig_latin.ex
defmodule PigLatin do
  @doc """
  Given a `phrase`, translate it a word at a time to Pig Latin.
  """
  @spec translate(phrase :: String.t()) :: String.t()
  def translate(phrase) do
    phrase
    |> String.split(" ")
    |> Enum.map(&translate_word/1)
    |> Enum.join(" ")
  end

  defp translate_word(word) do
    case word do
      <<char::binary-size(1), _::binary>> when char in ~w(a e i o u) ->
        word <> "ay"

      <<char1::binary-size(1), char2::binary-size(1), _::binary>>
      when char1 in ~w(x y) and char2 not in ~w(a e i o u y) ->
        word <> "ay"

      _ ->
        translate_word_default(word)
    end
  end

  defp translate_word_default(word) do
    case String.split(word, ~r/a|i|u|e|o|y|qu/, parts: 2, include_captures: true, trim: true) do
      [consonant, rest] ->
        rest <> consonant <> "ay"

      [prefix, vowel, rest] when vowel in ~w(a e i o u y) ->
        vowel <> rest <> prefix <> "ay"

      [prefix, "qu", rest] ->
        rest <> prefix <> "quay"
    end
  end
end
